onnx/defs/tensor/defs.cc

```diff
@@ -1185,4 +1185,122 @@ ONNX_OPERATOR_SET_SCHEMA(
+            "where 'on_value' is the value used for filling locations specified in 'indices' input "
+            "tensor, and 'off_value' is the value used for filling locations other than those specified "
+            "in 'indices' input tensor. ",
+            "T2")
+        .Output(0,
+            "output",
+            "Tensor of rank one greater than input tensor 'indices', i.e. rank(output) = rank(indices) + 1. "
+            "The data type for the elements of the output tensor is the same as the type of input 'values' "
+            "is used.",
+            "T2")
+        .TypeConstraint(
+            "T1",
+            OpSchema::all_numeric_types(),
+            "Constrains input to only numeric types.")
+        .TypeConstraint(
+            "T2",
+            OpSchema::all_tensor_types(),
+            "Constrain to any tensor type.")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          // Check that the node has two or three onputs.
+          if (!hasNInputShapes(ctx, 3)) {
+            fail_shape_inference(
+                  "OneHot node must have three inputs.");
+          }
+          // If input 'values' is specified, it must be a two-element vector.          
+          auto& input_shape = getInputShape(ctx, 2);
+          if (input_shape.dim_size() != 1) {
+            fail_type_inference(
+                  "Input 'values' must be rank 1 tensor.");
+          }
+          if (!input_shape.dim((int)0).has_dim_value() ||
+            (input_shape.dim((int)0).has_dim_value() &&
+              input_shape.dim((int)0).dim_value() != 2)) {
+                fail_type_inference(
+                      "Input 'values' must have exactly two elements.");
+          }
+          // Set output type to be the same as the third input, 'values'.
+          propagateElemTypeFromInputToOutput(ctx, 2, 0);
+          // Set the output shape
+          const TensorShapeProto& indices_shape = ctx.getInputType(0)->tensor_type().shape();
+          int r = indices_shape.dim_size();
+          if (r < 1) {
+            fail_shape_inference("Indices tensor must have rank >= 1");
+          }
+          int out_rank = r + 1;
+          int axis = static_cast<int>(getAttribute(ctx, "axis", -1));
+          if (axis < -out_rank || axis >= out_rank) {
+            fail_shape_inference("'axis' must be in [-rank(indices)-1, rank(indices)]");
+          }
+          if (axis < 0) {
+            axis += out_rank;
+          }
+          auto* output_shape = getOutputShape(ctx, 0);
+          for (int i = 0; i < out_rank; ++i) {
+            auto* dim = output_shape->add_dim();
+            if (i < axis) {
+              if (indices_shape.dim(i).has_dim_value()) {
+                dim->set_dim_value(indices_shape.dim(i).dim_value());
+              } else if (indices_shape.dim(i).has_dim_param()) {
+                dim->set_dim_param(indices_shape.dim(i).dim_param());
+              }
+            }
```

Comment:
how about i == axis case?

Reply:
In the `i == axis` case the size of that dimension comes from the `depth` input, which may not be known or inferrable. Therefore, for that case, we just add a dimension but do not add a value. Also, added test points for this scenario.